pubtools/pulplib/_impl/client/search.py

```diff
@@ -87,4 +89,12 @@ def field_match(to_match):
     if isinstance(to_match, ExistsMatcher):
         return {"$exists": True}
 
+    if isinstance(to_match, LessThanMatcher):
+        value = to_match._value
+        # only value of the format YYYY-mm-ddTHH:MM:SSZ is
+        # treated as date
+        if _is_iso_date_format(value):
```

Comment:
Interesting.  I'm not sure this is right, because we have some fields stored in Pulp which are genuinely stored as strings in the ISO8601 format, and not stored as dates.  Seems like this might make them unsearchable with LessThan, as there's no way to opt-out of using $date operator here if the value looks like ISO8601 format.

I think notes.created is an example of such a field.

Question: can we just base this on the type of the value instead?  I mean if _value is a datetime.datetime, then it's appropriate to search with "$date" operator.  Otherwise it's not.  I don't like the library "guessing" that a passed string should be interpreted as a date.

Reply:
I did some playing around with this locally and found:

- it's pretty easy to make it accept datetime.datetime objects from the caller. 
 Quick and dirty example: https://gist.github.com/rohanpm/88931ee34b3c0c4cefed0bd933766af2
- it's also possible to support searching by *both* datetime and str if we like.  It's not included in above gist.  It can be done by taking the above change, and then adding a bit more to support converting a str into a datetime, around the bit of the code where PY_PULP2_CONVERTER is currently accessed.

Reply:
I think using datetime.datetime object as input for the matcher will surely cut down the guessing for date. Wasn't sure if that's an option.

Agree that matcher._value type is used as the source of truth for the field type to get the desired result from the query.

So when you mention that both str and datetime can be used, should there be a check whether the type of the model field and matcher._value is same. If not, then convert it to the matcher._value type?

Reply:
I am not sure that we actually need the str => datetime conversion.  Guess it would be a nice convenience, but you could consider first getting a patch without that and then letting it be added later.

If added, yes I think it'd work pretty similar to the way you just described.  Look up some converter function based on the model field's declared type, then try to apply it on the matcher._value.